Add support for open.default.remote config

ffes opened the pull request:
Normally `git open` opens the remote named `origin`. When you fork a project and add a remote named `upstream` you often want that upstream to be opened as the default remote. This introduces a new `git config` setting to fulfill that need.

derimagia commented:
Nice, I was wanting something like this. What's the reasoning for not just going to the tracked remote?

ffes commented:
Not sure what the exact consequences of that would be.

What would be the remote after I push a feature branch to my fork `origin` to work / create a pull request? I still most likely want to open `upstream`. So a setting like this is more predictable. But I could be wrong.

derimagia commented:
Right now it defaults to "origin" anyway, which to me is the expected case. Maybe the "open pr" command would go to open upstream by default if it exists, though. I'm personally not for or against settings for this, though.

ffes commented:
The default still is `origin`. This PR is useful when the default doesn't work for you.

For instance, when I develop for *this* project, I use the normal github work-flow: My own fork with the remote origin and the upstream with that name. `origin` is only used to push feature branches to. Everything I need to lookup (issues, PRs, the wiki, etc) are on `upstream`. `upstream` is where this project is. So when I am in that local working directory I always need to do `git open upstream`. To make my life easier I created this PR because I keep forgetting to add the remote. Once set, I know that just `git open` will take me to the project's webpage, no matter what working directory I'm in.

paulirish commented:
@ffes you said

> So I think I will when this PR is merged I will update #79 as follows: 
> 1. remote to origin
> 2. if there is a tracked remote use that (#88)
> 3. if `gitopen.remote.default` is set use that (#79)
> 4. if `$1` is set use that. The user must always be able to override the magic 

that sgtm. seems like a good solution.

thanks

ffes commented:
SGTM as well

ffes commented:
I have written the new code to update this PR. Still need to update my branch with the current master and write the tests.

ffes commented:
Another thing that I noticed is that the GitLab configuration options have been changed from `gitopen.gitlab.*` to `open.*`. So I renamed this option to `open.default.remote` and updated the title of this PR.

paulirish commented:
looks good. I filed #106 to track tests for this.

derimagia commented:
Little of a chicken and egg thing but also: Would like to see if it's possible to add `--get-urlmatch` to this like custom domains have. Maybe just passing it the server or something? I'll try to get a chance to look at it

This way people could have the default branch on their company github enterprise be different than another git site globally if the workflow is different for example.
